pkg/kubelet/kubelet.go

```diff
@@ -2187,11 +2187,12 @@ func (kl *Kubelet) syncLoopIteration(configCh <-chan kubetypes.PodUpdate, handle
 			}
 			glog.V(2).Infof("SyncLoop (PLEG): %q, event: %#v", format.Pod(pod), e)
 			handler.HandlePodSyncs([]*api.Pod{pod})
-		}
-		if e.Type == pleg.ContainerDied {
-			if podStatus, err := kl.podCache.Get(e.ID); err == nil {
-				if containerID, ok := e.Data.(string); ok {
-					kl.containerDeletor.deleteContainersInPod(containerID, podStatus)
+
+			if e.Type == pleg.ContainerDied {
+				if podStatus, err := kl.podCache.Get(e.ID); err == nil {
+					if containerID, ok := e.Data.(string); ok {
+						kl.containerDeletor.deleteContainersInPod(containerID, podStatus, kl.podIsTerminated(pod))
```

Comment:
Once the pod is terminated, we'll never do sync pod, which means that:
1) We'll never try to restart a container.
2) We'll never generate new api pod status.

So, it's safe to delete container here.

1) For pod gracefully deleted, it usually terminates after all containers are dead. The "remove all" will almost never happen. The containers left will be removed by garbage collector.
2) For pod forcefully deleted, the pod will be removed from the pod manager, so the "remove" will never happen. The containers left will be removed by garbage collector.
3) For past active deadline and eviction, pod is set to terminated before killing containers. The deletion will be triggered for each dead container.

This is good enough, because the immediate removing is mainly for pod eviction.
A possible optimization is to not rely on pod manager, just get the pod uid from podStatus, and get status from the apiserver, so that we can handle case 2).

@ronnielai 
/cc @yujuhong to make sure I understand it correctly.